Applications/Utils/ModelPreparation/PartitionMesh/NodeWiseMeshPartitioner.cpp

```diff
@@ -84,106 +85,195 @@ void NodeWiseMeshPartitioner::readMetisData(const std::string& file_name_base)
     std::remove(fname_eparts.c_str());
 }
 
-void NodeWiseMeshPartitioner::partitionByMETIS(
-    const bool is_mixed_high_order_linear_elems)
+void NodeWiseMeshPartitioner::findNonGhostNodesInPartition(
+    std::size_t part_id,
+    const bool is_mixed_high_order_linear_elems,
+    std::vector<MeshLib::Node*>& extra_nodes)
 {
     std::vector<MeshLib::Node*> const& nodes = _mesh->getNodes();
-    for (std::size_t part_id = 0; part_id < _partitions.size(); part_id++)
+    auto& partition = _partitions[part_id];
+    // -- Extra nodes for high order elements
+    for (std::size_t i = 0; i < _mesh->getNumberOfNodes(); i++)
     {
-        auto& partition = _partitions[part_id];
+        if (_nodes_partition_ids[i] == part_id)
+        {
+            if (is_mixed_high_order_linear_elems)
+            {  // TODO: Test it once there is a case
+                if (i < _mesh->getNumberOfBaseNodes())
+                    partition.nodes.push_back(nodes[i]);
+                else
+                    extra_nodes.push_back(nodes[i]);
+            }
+            else
+            {
+                partition.nodes.push_back(nodes[i]);
+            }
+        }
+    }
+    partition.number_of_non_ghost_base_nodes = partition.nodes.size();
+    partition.number_of_non_ghost_nodes =
+        partition.number_of_non_ghost_base_nodes + extra_nodes.size();
+}
 
-        INFO("Processing partition: %d", part_id);
+void NodeWiseMeshPartitioner::findElementsInPartition(
+    std::size_t part_id, const bool is_mixed_high_order_linear_elems)
+{
+    auto& partition = _partitions[part_id];
+    std::vector<MeshLib::Element*> const& elements = _mesh->getElements();
+    for (std::size_t elem_id = 0; elem_id < elements.size(); elem_id++)
+    {
+        const auto* elem = elements[elem_id];
+        if (_elements_status[elem_id])
```

Comment:
I'm unsure but I think the vector `_element_status` contains a flag for every element. If the flag is set to true the element is a regular element (an element without ghost nodes), else it is set to false. Maybe a renaming of the vector raises the readability, too? Maybe @wenqing can confirm this?!